src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorLanguageEndpoint.cs

```diff
@@ -298,5 +298,96 @@ public async Task<RazorMapToDocumentEditsResponse> Handle(RazorMapToDocumentEdit
                 HostDocumentVersion = documentVersion,
             };
         }
+
+        public async Task<RazorDiagnosticsResponse> Handle(RazorDiagnosticsParams request, CancellationToken cancellationToken)
+        {
+            if (request is null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            var unmappedDiagnostics = request.Diagnostics;
+            var filteredDiagnostics = unmappedDiagnostics.Where(d => !CanDiagnosticBeFiltered(d));
+            if (!filteredDiagnostics.Any())
+            {
+                // No diagnostics left after filtering.
+                return new RazorDiagnosticsResponse()
+                {
+                    Diagnostics = Array.Empty<Diagnostic>()
+                };
+            }
+
+            var rangesToMap = filteredDiagnostics.Select(r => r.Range).ToArray();
+            var mappingParams = new RazorMapToDocumentRangesParams()
+            {
+                Kind = request.Kind,
+                RazorDocumentUri = request.RazorDocumentUri,
+                ProjectedRanges = rangesToMap,
+                MappingBehavior = request.MappingBehavior
+            };
+            var mappingResult = await Handle(mappingParams, cancellationToken).ConfigureAwait(false);
+
+            if (mappingResult == null || mappingResult.HostDocumentVersion != request.HostDocumentVersion)
```

Comment:
It'd make more sense to respond with the mapped document version and let the client validate that the response is compatible with the request. Aka, don't have a HostDocumentVersion on the request. That'd be more consistent with our other endpoints